icu4c/source/i18n/number_fluent.cpp

```diff
@@ -451,15 +462,40 @@ void LocalizedNumberFormatter::lnfMoveHelper(LNF&& src) {
     umtx_storeRelease(*callCount, INT32_MIN);
     delete fCompiled;
     fCompiled = src.fCompiled;
+    delete fWarehouse;
+    fWarehouse = src.fWarehouse;
     // Reset the source object to leave it in a safe state.
     auto* srcCallCount = reinterpret_cast<u_atomic_int32_t*>(src.fUnsafeCallCount);
     umtx_storeRelease(*srcCallCount, 0);
     src.fCompiled = nullptr;
+    src.fWarehouse = nullptr;
+}
+
+void LocalizedNumberFormatter::lnfCopyHelper(const LNF&, UErrorCode& status) {
+    // If MacroProps has a reference to AffixPatternProvider, we need to copy it.
+    // If MacroProps has a reference to PluralRules, copy that one, too.
+    if (fMacros.affixProvider || fMacros.rules) {
+        auto warehouse = new DecimalFormatWarehouse();
+        if (!warehouse) {
+            status = U_MEMORY_ALLOCATION_ERROR;
+            return;
+        }
+        if (fMacros.affixProvider) {
+            warehouse->affixProvider.setTo(fMacros.affixProvider, status);
+            fMacros.affixProvider = &warehouse->affixProvider.get();
+        }
+        if (fMacros.rules) {
+            warehouse->rules.adoptInstead(new PluralRules(*fMacros.rules));
```

Comment:
Maybe `adoptInsteadAndCheckErrorCode` to set the error to `U_MEMORY_ALLOCATION_ERROR` if `new` fails to allocate the PluralRules.

Reply:
Done